src/librustc_mir/dataflow/mod.rs

```diff
@@ -380,6 +380,62 @@ pub fn state_for_location<'tcx, T: BitDenotation<'tcx>>(loc: Location,
     gen_set.to_dense()
 }
 
+/// Calls `f` with the dataflow state at every location in `mir`.
+/// Ignores blocks that terminate in `unreachable`.
+pub fn for_each_location<'tcx, T: BitDenotation<'tcx>>(
+    mir: &Mir<'tcx>,
+    analysis: &T,
+    result: &DataflowResults<'tcx, T>,
+    mut f: impl FnMut(&HybridBitSet<T::Idx>, Location)
+) {
+    for (block, bb_data) in mir.basic_blocks().iter_enumerated() {
+        if bb_data.is_unreachable() {
+            continue;
```

Comment:
I'm confused as to why it matters if you have a block that has no statements and an `Unreachable`.

Reply:
Many edges might point to the `Unreachable` block, with different storage liveness for different vars. We shouldn't count all of those as conflicts at the beginning of the block.

Reply:
If there's nothing else in the block, can't you just ignore `Location`s pointing to an `Unreachable` terminator, in the user of this infrastructure, and not special-case it here?

Reply:
I certainly can, but it seemed like something more users would want. I don't have a strong opinion here.

I think if we use RPO we should keep the special-casing and rename this to `for_each_reachable_location` or something like that. Otherwise, we put the special casing in the user code. Does that sound good?

Reply:
Well, RPO would still visit the `Unreachable` block (since it's not *itself* statically unreachable).
So it wouldn't solve *that* part.
I guess I'm fine with leaving this as-is and moving the special-casing to the user code.

> it seemed like something more users would want

Usually dataflow is used by inspecting the state while looking at an `Operand`, `Rvalue` etc.
E.g. disallowing accesses to variables that aren't fully initialized (known via dataflow state).
So they wouldn't be affected by this.

Reply:
Right, I was thinking we could include RPO _and_ the `Unreachable` special-casing as one utility. But it sounds like maybe there would be cases where you would want RPO but not the special casing.

In any case, I don't know how to implement RPO yet and would rather not include that in this change :)

Reply:
You don't "implement RPO", you use it like pretty much any pass that cares about control/data-flow.
Like, it's the "standard and correct-by-default" traversal method for MIR.